# Segmenting and Clustering Neighborhoods in Toronto

_The first task requires to scrape and parse the given web-page. I'm going to use BeautifulSoup4 and requests libraries for this purpose._

In [1]:
# using requests load given web-page into object page 
import requests
page = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
page

<Response [200]>

_Also I need pandas for loading parsed data into dataframe_

In [2]:
import pandas as pd

_Let's install BeautifulSoup4_

In [ ]:
!pip install beautifulsoup4

In [ ]:
!conda install -c anaconda beautifulsoup4

In [3]:
#here we parse page content
page.content
from bs4 import BeautifulSoup
soup = BeautifulSoup(page.content, 'html.parser')

In [ ]:
#print(soup.prettify())

_Let's create dataframe for our data. It consists of 3 columns PostalCode, Borough, Neighborhood. And than load data into it from object soup_ 

In [4]:
#Let's create dataframe for our data. It consists of 3 columns  
CACodeDf = pd.DataFrame(columns=["PostalCode", "Borough", "Neighborhood"])

In [8]:
#And finaly, let's load data from our page into dataframe  CACodeDf
i = 0
for tr in soup.find_all('tr', text=True)[1:]:
    tds = tr.find_all('td')  
    my_str = tds[2].text.replace('\n', '')
    CACodeDf.loc[i] = [tds[0].text]+ [tds[1].text] + [my_str]
    i = i+1
    #print(tds[0].text)
print      ("done")

done


_Here is the result. First 5 rows of our dataset_

In [9]:
CACodeDf.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [ ]:
CACodeDf.shape

_Now I have to process the data accordingly instructions given into assignment. First, I should remove all rows where Borough is Not assigned_

In [10]:
#Now , since we have dataframe filled with all necessary data, we should clear it from inappropriate rows
CACodeDf = CACodeDf[CACodeDf.Borough != "Not assigned"]
CACodeDf.head() 

PostalCode           Borough      Neighborhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M5A  Downtown Toronto       Regent Park
6        M6A        North York  Lawrence Heights

In [ ]:
CACodeDf.shape

_Now, let's update values of the cells where  Neighborhood is Not assigned. The new value of Neighborhood  for such rows must be equal its Borough value_

In [11]:
CACodeDf.loc[CACodeDf['Neighborhood'] == "Not assigned", ['Neighborhood']] = CACodeDf['Borough']
CACodeDf.head()

PostalCode           Borough      Neighborhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M5A  Downtown Toronto       Regent Park
6        M6A        North York  Lawrence Heights

In [ ]:
CACodeDf.shape

_Finally, I have to transform our dataframe into the new one, so Postal Code would be unique_

In [12]:
df = CACodeDf[['PostalCode','Borough']]
#df.drop_duplicates()
df=df.drop_duplicates(['PostalCode','Borough'])[['PostalCode','Borough']]
df['Neighborhood'] = df.apply(lambda row: "", axis=1)
df.head()
#df.shape

PostalCode           Borough Neighborhood
2        M3A        North York             
3        M4A        North York             
4        M5A  Downtown Toronto             
6        M6A        North York             
8        M7A      Queen's Park

In [14]:
for index, row in df.iterrows():
    my_str = ''
    for index2, row2 in CACodeDf.iterrows():
        if row2['PostalCode'] == row['PostalCode'] and row2['Borough'] == row['Borough']:
            my_str = row2['Neighborhood']+","+my_str
    row['Neighborhood'] = my_str[:-1]
df.reset_index(inplace=True)    
df.drop('index',axis=1, inplace=True)
df.head()     

PostalCode           Borough                     Neighborhood
0        M3A        North York                        Parkwoods
1        M4A        North York                 Victoria Village
2        M5A  Downtown Toronto         Regent Park,Harbourfront
3        M6A        North York  Lawrence Manor,Lawrence Heights
4        M7A      Queen's Park                     Queen's Park

In [15]:
df.shape

(103, 3)

_So, you can see the final dataset consists of 103 rows.<br> The next step is defining location for each Borough. I prefer using the csv of Geospatial data. And the line below imports the csv into my project_

In [ ]:
!wget -O Geospatial_Coordinates.csv https://cocl.us/Geospatial_data

_Now let's load location data into the new dataftame  dfcoord_

In [18]:
df_coord = pd.read_csv('Geospatial_Coordinates.csv')
df_coord.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

_Let's increase our dataframe df with 2 columns Latitude and Longitude and assign its values from the dataframe dfcoord shown above_

In [19]:
df['Latitude'] = df.apply(lambda row: "", axis=1)
df['Longitude'] = df.apply(lambda row: "", axis=1)
df.head() 

PostalCode           Borough                     Neighborhood Latitude  \
0        M3A        North York                        Parkwoods            
1        M4A        North York                 Victoria Village            
2        M5A  Downtown Toronto         Regent Park,Harbourfront            
3        M6A        North York  Lawrence Manor,Lawrence Heights            
4        M7A      Queen's Park                     Queen's Park            

  Longitude  
0            
1            
2            
3            
4

In [20]:
for index, row in df.iterrows():
    for index2, row2 in df_coord.iterrows():
        if row2['Postal Code'] == row['PostalCode'] :            
            row['Latitude'] = row2['Latitude']
            row['Longitude'] = row2['Longitude']
#df.reset_index(inplace=True)
#df.drop('index',axis=1, inplace=True)
df.head(10)    

PostalCode           Borough                     Neighborhood Latitude  \
0        M3A        North York                        Parkwoods  43.7533   
1        M4A        North York                 Victoria Village  43.7259   
2        M5A  Downtown Toronto         Regent Park,Harbourfront  43.6543   
3        M6A        North York  Lawrence Manor,Lawrence Heights  43.7185   
4        M7A      Queen's Park                     Queen's Park  43.6623   
5        M9A         Etobicoke                 Islington Avenue  43.6679   
6        M1B       Scarborough                    Malvern,Rouge  43.8067   
7        M3B        North York                  Don Mills North  43.7459   
8        M4B         East York   Parkview Hill,Woodbine Gardens  43.7064   
9        M5B  Downtown Toronto          Garden District,Ryerson  43.6572   

  Longitude  
0  -79.3297  
1  -79.3156  
2  -79.3606  
3  -79.4648  
4  -79.3895  
5  -79.5322  
6  -79.1944  
7  -79.3522  
8  -79.3099  
9  -79.3789

_Here is the final dataset ready for further exploration_

In [21]:
df.shape

(103, 5)

In [ ]:
df.dtypes

_Now I define a new dataset , subset of df data, where I include only Boroughs which contain "Toronto" in its name_

In [22]:
toronto_data = df[df['Borough'].str.contains("Toronto") == True].reset_index(drop=True)
toronto_data.head()

PostalCode           Borough              Neighborhood Latitude Longitude
0        M5A  Downtown Toronto  Regent Park,Harbourfront  43.6543  -79.3606
1        M5B  Downtown Toronto   Garden District,Ryerson  43.6572  -79.3789
2        M5C  Downtown Toronto            St. James Town  43.6515  -79.3754
3        M4E      East Toronto               The Beaches  43.6764   -79.293
4        M5E  Downtown Toronto               Berczy Park  43.6448  -79.3733

In [23]:
toronto_data.shape

(38, 5)

### Clusterisation and exploration Neighborhoods in Toronto

_Before we begin data exploration we have to download  libraries and dependencies we will need_

In [ ]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

_Let's define location and coordinates for Toronto city_

In [25]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.653963, -79.387207.


_Let's drow a map based on whole dataset df_

In [26]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, df in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(df, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

_And now, let's take a look at a map based on torontodata dataset_

In [27]:
# create map of Toronto using latitude and longitude values
map_toronto1 = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto1)  
    
map_toronto1

_For further utilizing the Foursquare API to explore the neighborhoods and segment them, let's define credentials_

In [28]:
CLIENT_ID = 'P5FRQ54L5KSCZASQMXUSHFQT2ZO2DK10SSPDSFPJE14ZWPTI' # your Foursquare ID
CLIENT_SECRET = 'L4ISFZWLMOHIEZR5V4TQ4HKFCOHESAUYRSHF1SAWJUTXAVHX' # your Foursquare Secret
VERSION = '20190630' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: P5FRQ54L5KSCZASQMXUSHFQT2ZO2DK10SSPDSFPJE14ZWPTI
CLIENT_SECRET:L4ISFZWLMOHIEZR5V4TQ4HKFCOHESAUYRSHF1SAWJUTXAVHX


_Just for example, let's see what Foursquare API would retrieve for the first element of our dataframe torontodata_

In [29]:
toronto_data.loc[0, 'Neighborhood']

'Regent Park,Harbourfront'

In [30]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park,Harbourfront are 43.6542599, -79.3606359.


In [31]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=P5FRQ54L5KSCZASQMXUSHFQT2ZO2DK10SSPDSFPJE14ZWPTI&client_secret=L4ISFZWLMOHIEZR5V4TQ4HKFCOHESAUYRSHF1SAWJUTXAVHX&v=20190630&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [ ]:
results = requests.get(url).json()
results

In [ ]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [126]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(10)

name            categories  \
0                                 Roselle Desserts                Bakery   
1                                    Tandem Coffee           Coffee Shop   
2  Toronto Cooper Koo Family Cherry St YMCA Centre  Gym / Fitness Center   
3                              Body Blitz Spa East                   Spa   
4                               Morning Glory Cafe        Breakfast Spot   
5                                   Impact Kitchen            Restaurant   
6                           Figs Breakfast & Lunch        Breakfast Spot   
7                         Dominion Pub and Kitchen                   Pub   
8                                  Corktown Common                  Park   
9                 The Distillery Historic District         Historic Site   

         lat        lng  
0  43.653447 -79.362017  
1  43.653559 -79.361809  
2  43.653191 -79.357947  
3  43.654735 -79.359874  
4  43.653947 -79.361149  
5  43.656369 -79.356980  
6  43.655675 -79.364503  
7  43.656919 -79.358967  
8  43.655618 -79.356211  
9  43.650244 -79.359323

_And how many venues were returned by Foursquare?_

In [125]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


### Explore Neighborhoods inToronto

_Now let's define a function for retrieving data using Foursquare API for each Neighborhood from our dataset_

In [32]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

_Now write the code to run the above function on each neighborhood and create a new dataframe called *torontovenues*_

In [33]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park,Harbourfront
Garden District,Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond,King,Adelaide
Dufferin,Dovercourt Village
Union Station,Toronto Islands,Harbourfront East
Trinity,Little Portugal
Riverdale,The Danforth West
Toronto Dominion Centre,Design Exchange
Parkdale Village,Exhibition Place,Brockton
India Bazaar,The Beaches West
Victoria Hotel,Commerce Court
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill West,Forest Hill North
The Junction South,High Park
North Toronto West
Yorkville,North Midtown,The Annex
Roncesvalles,Parkdale
Davisville
University of Toronto,Harbord
Swansea,Runnymede
Summerhill East,Moore Park
Kensington Market,Grange Park,Chinatown
Summerhill West,South Hill,Rathnelly,Forest Hill SE,Deer Park
South Niagara,Railway Lands,King and Spadina,Harbourfront West,Island airport,Bathurst Quay,CN Tower
Rosedale
Stn A PO Boxes 25 The Esplanade
St. James Town,Cabbagetown
Underground city,First Canadian Place


_Resulting dataframe is shown below_

In [34]:
print(toronto_venues.shape)
toronto_venues.head()

(1701, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park,Harbourfront               43.65426              -79.360636   
1  Regent Park,Harbourfront               43.65426              -79.360636   
2  Regent Park,Harbourfront               43.65426              -79.360636   
3  Regent Park,Harbourfront               43.65426              -79.360636   
4  Regent Park,Harbourfront               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                              Body Blitz Spa East       43.654735   
4                               Morning Glory Cafe       43.653947   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.359874                   Spa  
4       -79.361149        Breakfast Spot

_Let's group all Neighborhoods by count of venues we obtained from the Foursquare API_

In [35]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Business Reply Mail Processing Centre 969 Eastern                      19   
Central Bay Street                                                     88   
Christie                                                               15   
Church and Wellesley                                                   87   
Davisville                                                             38   
Davisville North                                                        9   
Dufferin,Dovercourt Village                                            20   
Forest Hill West,Forest Hill North                                      4   
Garden District,Ryerson                                               100   
India Bazaar,The Beaches West                                          18   
Kensington Market,Grange Park,Chinatown                               100   
Lawrence Park                                                           3   
North Toronto West                                                     16   
Parkdale Village,Exhibition Place,Brockton                             22   
Regent Park,Harbourfront                                               48   
Richmond,King,Adelaide                                                100   
Riverdale,The Danforth West                                            42   
Roncesvalles,Parkdale                                                  15   
Rosedale                                                                5   
Roselawn                                                                2   
South Niagara,Railway Lands,King and Spadina,Ha...                     16   
St. James Town                                                        100   
St. James Town,Cabbagetown                                             46   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        38   
Summerhill East,Moore Park                                              2   
Summerhill West,South Hill,Rathnelly,Forest Hil...                     14   
Swansea,Runnymede                                                      38   
The Beaches                                                             5   
The Junction South,High Park                                           23   
Toronto Dominion Centre,Design Exchange                               100   
Trinity,Little Portugal                                                62   
Underground city,First Canadian Place                                 100   
Union Station,Toronto Islands,Harbourfront East                       100   
University of Toronto,Harbord                                          34   
Victoria Hotel,Commerce Court                                         100   
Yorkville,North Midtown,The Annex                                      23   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Business Reply Mail Processing Centre 969 Eastern                       19   
Central Bay Street                                                      88   
Christie                                                                15   
Church and Wellesley                                                    87   
Davisville                                                              38   
Davisville North                                                         9   
Dufferin,Dovercourt Village                                             20   
Forest Hill West,Forest Hill North                                       4   
Garden District,Ryerson                                                100   
India Bazaar,The Beaches West      

In [36]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 240 uniques categories.


### Analyze Each Neighborhood

In [37]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Auto Workshop  \
0           0                    0                 0              0   
1           0                    0                 0              0   
2           0                    0                 0              0   
3           0                    0                 0              0   
4           0                    0                 0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               1        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Came

In [38]:
toronto_onehot.shape

(1701, 240)

In [42]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Berczy Park     0.000000   
1  Business Reply Mail Processing Centre 969 Eastern     0.052632   
2                                 Central Bay Street     0.011364   
3                                           Christie     0.000000   
4                               Church and Wellesley     0.011494   

   Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                0.0        0.000000           0.000000      0.0   
1                0.0        0.000000           0.000000      0.0   
2                0.0        0.000000           0.000000      0.0   
3                0.0        0.000000           0.000000      0.0   
4                0.0        0.011494           0.011494      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0               0.0             0.000000           0.0       0.0     0.018182   
1               0.0             0.000000           0.0       0.0     0.000000   
2               0.0             0.011364           0.0       0.0     0.000000   
3               0.0             0.000000           0.0       0.0     0.000000   
4               0.0             0.011494           0.0       0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Auto Workshop  \
0    0.000000             0.000000               0.0       0.000000   
1    0.000000             0.000000               0.0       0.052632   
2    0.011364             0.011364               0.0       0.000000   
3    0.000000             0.000000               0.0       0.000000   
4    0.000000             0.011494               0.0       0.000000   

   BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0        0.0    0.000000    0.018182  0.036364   0.0  0.000000   
1        0.0    0.000000    0.000000  0.000000   0.0  0.000000   
2        0.0    0.000000    0.000000  0.022727   0.0  0.022727   
3        0.0    0.066667    0.000000  0.000000   0.0  0.000000   
4        0.0    0.000000    0.000000  0.000000   0.0  0.000000   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.018182  0.018182              0.0  0.036364   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Boat or Ferry  Bookstore  \
0         0.0            0.018182  0.018182            0.0   0.000000   
1         0.0            0.000000  0.000000            0.0   0.000000   
2         0.0            0.000000  0.000000            0.0   0.000000   
3         0.0            0.000000  0.000000            0.0   0.000000   
4         0.0            0.000000  0.000000            0.0   0.011494   

   Boutique  Brazilian Restaurant  Breakfast Spot   Brewery  Bubble Tea Shop  \
0       0.0                   0.0        0.018182  0.000000         0.000000   
1       0.0                   0.0        0.000000  0.052632         0.000000   
2       0.0                   0.0        0.000000  0.000000         0.022727   
3       0.0                   0.0        0.000000  0.000000         0.000000   
4       0.0                   0.0        0.011494  0.000000         0.022989   

   Building  Burger Joint  Burrito Place  Bus Line   Butcher      Café 

In [41]:
toronto_grouped.shape

(38, 240)

In [43]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2  Italian Restaurant  0.04
3         Cheese Shop  0.04
4      Farmers Market  0.04


----Business Reply Mail Processing Centre 969 Eastern----
              venue  freq
0       Yoga Studio  0.05
1     Auto Workshop  0.05
2        Comic Shop  0.05
3              Park  0.05
4  Recording Studio  0.05


----Central Bay Street----
                       venue  freq
0                Coffee Shop  0.16
1         Italian Restaurant  0.05
2                       Café  0.05
3               Burger Joint  0.03
4  Middle Eastern Restaurant  0.03


----Christie----
               venue  freq
0               Café  0.20
1      Grocery Store  0.20
2               Park  0.13
3         Baby Store  0.07
4  Convenience Store  0.07


----Church and Wellesley----
                 venue  freq
0  Japanese Restaurant  0.07
1          Coffee Shop  0.07
2     Sushi Restaurant  0.06
3           Restaurant  0.03


In [45]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [46]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie                  Café   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Italian Restaurant           Cheese Shop   
1      Recording Studio            Smoke Shop            Skate Park   
2                  Café    Italian Restaurant          Burger Joint   
3         Grocery Store                  Park    Italian Restaurant   
4   Japanese Restaurant      Sushi Restaurant               Gay Bar   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                       Café        Farmers Market            Steakhouse   
1                    Brewery         Burrito Place               Butcher   
2  Middle Eastern Restaurant        Sandwich Place       Bubble Tea Shop   
3                      Diner            Restaurant             Nightclub   
4                 Restaurant                 Hotel       Bubble Tea Shop   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Bakery    Seafood Restaurant               Beer Bar  
1                Restaurant                  Park          Auto Workshop  
2                       Spa      Sushi Restaurant     Chinese Restaurant  
3                Baby Store           Coffee Shop      Convenience Store  
4  Mediterranean Restaurant                  Café           Burger Joint

 ### Cluster Neighborhoods  

_Run *k*-means to cluster the neighborhood into 5 clusters._

In [47]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 4, 0], dtype=int32)

In [48]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough              Neighborhood Latitude Longitude  \
0        M5A  Downtown Toronto  Regent Park,Harbourfront  43.6543  -79.3606   
1        M5B  Downtown Toronto   Garden District,Ryerson  43.6572  -79.3789   
2        M5C  Downtown Toronto            St. James Town  43.6515  -79.3754   
3        M4E      East Toronto               The Beaches  43.6764   -79.293   
4        M5E  Downtown Toronto               Berczy Park  43.6448  -79.3733   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop                   Pub   
1               0           Coffee Shop        Clothing Store   
2               0           Coffee Shop                  Café   
3               0     Health Food Store  Other Great Outdoors   
4               0           Coffee Shop          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                Bakery                  Park         Mexican Restaurant   
1                  Café        Cosmetics Shop  Middle Eastern Restaurant   
2                 Hotel            Restaurant             Cosmetics Shop   
3                   Pub                 Trail                Event Space   
4    Italian Restaurant           Cheese Shop                       Café   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Theater            Restaurant        Breakfast Spot   
1                 Diner       Bubble Tea Shop           Pizza Place   
2        Breakfast Spot                Bakery             Gastropub   
3  Ethiopian Restaurant    Falafel Restaurant     Electronics Store   
4        Farmers Market            Steakhouse                Bakery   

  9th Most Common Venue 10th Most Common Venue  
0                  Café   Gym / Fitness Center  
1    Italian Restaurant    Japanese Restaurant  
2    Italian Restaurant                  Diner  
3          Dessert Shop         Farmers Market  
4    Seafood Restaurant               Beer Bar

In [49]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [50]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3       East Toronto               0     Health Food Store   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0                  Café   
7   Downtown Toronto               0           Coffee Shop   
8       West Toronto               0              Pharmacy   
9   Downtown Toronto               0           Coffee Shop   
10      West Toronto               0                   Bar   
11      East Toronto               0      Greek Restaurant   
12  Downtown Toronto               0           Coffee Shop   
13      West Toronto               0           Coffee Shop   
14      East Toronto               0                  Park   
15  Downtown Toronto               0           Coffee Shop   
16      East Toronto               0                  Café   
19   Central Toronto               0            Playground   
21      West Toronto               0                  Café   
22   Central Toronto               0           Coffee Shop   
23   Central Toronto               0           Coffee Shop   
24      West Toronto               0        Breakfast Spot   
25   Central Toronto               0           Pizza Place   
26  Downtown Toronto               0                  Café   
27      West Toronto               0           Coffee Shop   
29  Downtown Toronto               0                  Café   
30   Central Toronto               0           Coffee Shop   
31  Downtown Toronto               0       Airport Service   
33  Downtown Toronto               0           Coffee Shop   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
37      East Toronto               0           Yoga Studio   

            2nd Most Common Venue 3rd Most Common Venue  \
0                             Pub                Bakery   
1                  Clothing Store                  Café   
2                            Café                 Hotel   
3            Other Great Outdoors                   Pub   
4                    Cocktail Bar    Italian Restaurant   
5                            Café    Italian Restaurant   
6                   Grocery Store                  Park   
7                            Café       Thai Restaurant   
8                     Supermarket                Bakery   
9                           Hotel              Aquarium   
10                    Men's Store      Asian Restaurant   
11                    Coffee Shop        Ice Cream Shop   
12                           Café                 Hotel   
13                 Breakfast Spot                  Café   
14                    Pizza Place      Sushi Restaurant   
15                           Café                 Hotel   
16                    Coffee Shop             Gastropub   
19                            Gym         Grocery Store   
21             Mexican Restaurant                   Bar   
22                    Yoga Studio            Bagel Shop   
23                           Café        Sandwich Place   
24                      Gift Shop    Italian Restaurant   
25                   Dessert Shop        Sandwich Place   
26             Italian Restaurant   Japanese Restaurant   
27                           Café           Pizza Place   
29  Vegetarian / Vegan Restaurant    Mexican Restaurant   
30                            Pub   American Restaurant   
31               Airport Terminal        Airport Lounge   
33                           Café            Restaurant   
34                     Restaurant                Bakery   
35                           Café                 Hotel   
36            Ja

In [114]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               1        Ice Cream Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18                Garden         Women's Store                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18           Event Space  Ethiopian Restaurant      Electronics Store

In [115]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Central Toronto               2            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28          Tennis Court          Dessert Shop        Farmers Market   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
28     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

In [116]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17           Swim School              Bus Line         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                 Diner        Farmers Market    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17           Event Space  Ethiopian Restaurant      Electronics Store

In [117]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20   Central Toronto               4                 Trail   
32  Downtown Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20         Jewelry Store                  Park      Sushi Restaurant   
32            Playground                 Trail              Building   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20  Eastern European Restaurant               Dog Run      Doner Restaurant   
32                        Diner        Farmers Market    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20            Donut Shop   Dumpling Restaurant          Women's Store  
32           Event Space  Ethiopian Restaurant      Electronics Store